src/daemon/utils_ec2_meta.c

```diff
@@ -0,0 +1,234 @@
+
+#include "collectd.h"
+
+#include "plugin.h"
+#include "common.h"
+#include "utils_format_json.h"
+#include "utils_format_kairosdb.h"
+
+#include <curl/curl.h>
+
+#define CURL_TIMEOUT_MS	500	/*ms*/
+#define EC2_FIELD_SIZE	512
+#define MAX_HDR_LEN	256
+
+#define BASE_EC2_URL	"http://169.254.169.254/latest/meta-data/"
+
+struct ec2_meta_field {
+	char	buf[EC2_FIELD_SIZE];
+	size_t	filled;
+};
+
+struct ec2_meta {
+	struct ec2_meta_field *instance_id;
+	struct ec2_meta_field *instance_type;
+	struct ec2_meta_field *az;
+};
+
+static struct ec2_meta *_ec2_meta_saved;
+
+static size_t ec2_meta_callback(void *buf, size_t size,
+    size_t nmemb, void *user_data)
+{
+	size_t len = size * nmemb;
+
+	struct ec2_meta_field *field = (struct ec2_meta_field *)user_data;
+	if (field == NULL) {
+		return len;
+	}
+
+	if (field->filled + len >= sizeof(field->buf)) {
+		return (0);
+	}
+
+	memcpy(field->buf + field->filled, (char *)buf, len);
+	field->filled += len;
+	field->buf[field->filled] = '\0';
+
+	return len;
+}
+
+static int ec2_meta_is_success(CURL *curl, int status)
+{
+	long http_code = 0;
+
+	curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &http_code);
+	return (http_code == 200 && status != CURLE_ABORTED_BY_CALLBACK) ? 0 : -1;
+}
+
+static int ec2_meta_check_ec2(CURL *curl)
+{
+	char url[1024];
+
+	snprintf(url, sizeof(url), BASE_EC2_URL);
+
+	curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
+	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, ec2_meta_callback);
+	curl_easy_setopt(curl, CURLOPT_WRITEDATA, NULL);
```

Comment:
Yeah so in this case the callback function is just dropping the data on the floor (similar to piping to /dev/null). I found that if you don't have a write function, curl will echo output to standard output. This function is simply ensuring we get a 200 from the API, the output we drop.